README
NAME
    Fuse::Template - Mount a directory with templates.

VERSION
    0.0201

USAGE
    "sudo" might be optional.

  MOUNT
    Example SQLite:

     $ sudo fuse-template
       --root /path/to/templates \
       --mountpoint /path/to/mountpoint \
       --schema dbi:SQLite:path/to/db \
       ;

    Example MySQL:

     $ sudo fuse-template
       --root /path/to/templates \
       --mountpoint /path/to/mountpoint \
       --schema "dbi:mysql:database=MyDB;host=127.1;port=3006 user pass" \
       ;

    See DBI for possible "--schema" formats.

  UNMOUNT
    This needs to be done manually for now. (Patches are welcome!)

     sudo umount /path/to/mountpoint;

DESCRIPTION
    IMPORTANT! Version 0.02 of Fuse::Template breaks backward compat by
    using the new "v7" naming convention from DBIx::Class::Schema::Loader.
    Set the "SCHEMA_LOADER_BACKCOMPAT" environment variable, if you want the
    old behavior.

    The idea with this project is to auto-maintain password files, and other
    plain text files for different systems, with data from a database.

    See "examples/" in this distribution for templates.

    Example file system structure:

     root/               -> mountpooit/
     root/somefile       -> mountpoint/somefile
     root/foo.tt         -> mountpoint/foo
     root/bar/baz.txt.tt -> mountpoint/bar/baz.txt

    "somefile" is accessible directly at "mountpoint". Template files (.tt)
    are parsed and the output is accessible on mountpoint side. This is done
    using Template::Toolkit.

    Files in mountpoint are read-only for now.

TEMPLATES
    Here is an example template:

     root = [% self.root %]
     [% rs = schema.resultset('MyTable') %]
     [% rs = MyTable; # same as schema.resultset(...) %]
     [% WHILE (row = rs.next) # get next row in resultset %]
     col_name = [% row.col_name # retrieve column data %]
     [% END %]

    See DBIx::Class for information on how to use the schema object, and
    DBIx::Class::Resultset on how to select data and more...

    Available variables:

     root
     mountpoint
     mountopts
     self # this object
     schema # DBIx::Class object
     + all resultsetset/tables

    Resultset use this naming convention to convert tablenames:

        Table Name  | Moniker Name
        ---------------------------
        luser       | Luser
        luser_group | LuserGroup
        luser-opts  | LuserOpts

    See "moniker_map" in DBIx::Class::Schema::Loader for details

INSTALLATION
    On ubuntu:

        $ sudo aptitude install libfuse-perl
        $ wget -q http://xrl.us/cpanm | perl - --sudo Fuse::Template

    See App::cpanminus if you are curious about the "wget" command.

ATTRIBUTES
  root
     $object = $self->root;

    $object must be stringified to path where templates/files are located.
    $object must follow the same API as Fuse::Template::Root.

    Can be set in constructor as:

     root => \%constructor_args
     root => $path_to_root

    %constructor_args can override defaults, while $path_to_root will only
    override "path" attribute in root-object.

  mountpoint
     $path = $self->mountpoint;

    Path to where the filesystem should be mounted.

  mountopts
     $str = $self->mountopts;

    Example mount option: "allow_other"

  schema
     $dbic_object = $self->schema;

    Can be set in constructor as:

     schema => \%args
     schema => "$dsn $username $password"

    See Fuse::Template::Schema for details.

  debug
     $bool = $self->debug;

    Enable/disable debug output from Fuse. Default is disabled.

METHODS
  find_file
     $real_path = $self->find_file($virtual_path);

    Returns path to the file in root path.

  log
     $bool = $self->log($level => $format => @args);

AUTHOR
    Jan Henning Thorsen

LICENSE
    This library is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Fuse-Template

0.0201   Sat Jun 19 13:49:36 CEST 2010
       * Forgot to update Makefile.PL
       * Add auto install of bin/fuse-template

0.02     Sat Jun 19 13:37:00 CEST 2010
       * Fix dependencies
       * Add "ft" is alias for "self" in templates
       * Add more documentation
       * Change default values for gid() and uid()
       * Break backward compat: naming conventions for database results
         (tables) and columns default to "v7". (see
         http://search.cpan.org/dist/DBIx-Class-Schema-Loader/lib/DBIx/Class/Schema/Loader/Base.pm#naming
         for details)


0.01     Mon Jul 27 22:24:00 CET 2009
       * Add support for mount ro filesystem
       * Add support for templates
       * Add can read data from schema
         - Dynamic database schema load with DBIx::Class::Schema::Loader



lib/Fuse/Template/TT.pm
package Fuse::Template::TT;

=head1 NAME

Fuse::Template::TT - Read template files

=cut

use Moose;
use Template;

=head1 ATTRIBUTES

=head2 vars

Holds a hash ref with the variables which is available in the template.

=cut

has vars => (
    is => 'ro',
    isa => 'HashRef',
    default => sub { {} },
);

=head2 paths

Holds an array ref with paths to where L<Template::Toolkit> will look
for templates.

=cut

has paths => (
    is => 'ro',
    isa => 'ArrayRef',
    default => sub { [] },
);

has _template => (
    is => 'ro',
    isa => 'Object',
    lazy_build => 1,
    handles => [qw/process error/],
);

sub _build__template {
    return Template->new(
        INCLUDE_PATH => $_[0]->paths,
        EVAL_PERL => 0,
        TEMPLATE_EXTENSION => 'tt',
    );
}

=head1 METHODS

=head2 render

    $text = $self->render($vfile);

Will return the output from a processed template. The C<$vfile> points
to a template relative to one of the L</paths>. It cannot start with "/".

C<$text> will hold the error message if templat toolkit fail to render
the template.

=cut

sub render {
    my $self = shift;
    my $vfile = shift;
    my $text;
    
    $self->process($vfile, $self->vars, \$text);

    return $text ? $text : $self->error;
}

=head2 process

See L<Template>.

=head2 error

See L<Template>.

=head1 AUTHOR

See L<Fuse::Template>.

=cut

1;


lib/Fuse/Template/App.pm
package Fuse::Template::App;

=head1 NAME

Fuse::Template::App - Application role for fuse-template

=head1 SYNOPSIS

See C<bin/fuse-template> for example.

=cut

use Moose::Role;
use Fuse;

with 'MooseX::Getopt';

MooseX::Getopt::OptionTypeMap->add_option_type_to_map(
    Fuse::Template::Root::RootObject, '=s'
);
MooseX::Getopt::OptionTypeMap->add_option_type_to_map(
    Fuse::Template::Schema::Schema, '=s'
);

=head1 ATTRIBUTES

=head2 help

Provides "--help", which again prints a list of command line
options.

=cut

has help => (
    is => 'ro',
    isa => 'Bool',
    documentation => 'This help text',
);

=head2 man

Provides "--man", which will run perldoc on L<Fuse::Template>.

=cut

has man => (
    is => 'ro',
    isa => 'Bool',
    documentation => 'Read Fuse::Template manual',
);

=head1 METHODS

=head2 run

Set up objects and starts the L<Fuse> mainloop.

=cut

sub run {
    my $self = shift;
    my %callbacks;

    if($self->help) {
        exit 0;
    }
    elsif($self->man) {
        exec perldoc => grep { m[ Fuse/Template.pm$ ]x } values %INC;
    }

    unless($self->mountpoint and $self->root) {
        warn "Usage: fuse-template --help\n\n";
        return 1;
    }

    for my $method (Fuse::Template::Sys->meta->get_method_list) {
        $callbacks{$method} = sub { $self->$method(@_) };
    }

    $self->log(info => "Starting Fuse mainloop");

    Fuse::main(
        %callbacks,
        mountpoint => $self->mountpoint,
        debug      => $self->debug || 0,
        mountopts  => $self->mountopts || q(),
        threaded   => 0,
    );

    return 0;
}

sub _threads_enabled {
    return 1 if(grep { $_ eq 'threads.pm' } keys %INC);
    return 0;
}

=head1 AUTHOR

See L<Fuse::Template>.

=cut

1;


lib/Fuse/Template/Sys.pm
package Fuse::Template::Sys;

=head1 NAME

Fuse::Template::Sys - Moose role for system calls

=head1 DESCRIPTION

This role requires C<find_file()> and C<log()>.

Documentation is mainly copy/paste from L<Fuse>.

It should be possible to use this role in other projects as well.
I might even seperate it out as a standalone distribution some day.

=cut

use Moose::Role;
use Fuse::Template::Root qw/RootObject/;
use Fuse ();
use POSIX ();

requires qw/find_file log/;

=head1 ATTRIBUTES

=head2 root

Holds a L<Fuse::Template::Root> object.

=cut

has root => (
    is => 'ro',
    isa => RootObject,
    coerce => 1,
    documentation => "Location to template files",
    #required => 1,
);

=head1 METHODS

=head2 getattr

 @stat = $self->getattr($virtual_path);

Return value is the same format as from C<stat($file)>.

=cut

sub getattr {
    my $self  = shift;
    my $vfile = shift;
    my $file  = $self->find_file($vfile);
    my $root  = $self->root;

    $self->log(debug => "getattr(%s)", $file);

    return -&POSIX::ENOENT unless($file); # such file or directory

    if(length $vfile <= 1) {
        return (
            0,                #  0 device number
            0,                #  1 inode number
            (0040 << 9) + $root->mode, #  2 file mode 
            1,                #  3 number of hard links
            $root->uid,       #  4 user id
            $root->gid,       #  5 group id
            0,                #  6 device indentifier
            4096,             #  7 size
            $root->atime,     #  8 last acess time
            $root->mtime,     #  9 last modified time
            $root->ctime,     # 10 last change time
            $root->block_size,# 11 block size
            0,                # 12 blocks
        );
    }
    else {
        my @stat = lstat $file;
        return -$! unless @stat;
        return @stat;
    }
}

=head2 readlink

 $bool = $self->readlink($virtual_path); 

This is called when dereferencing symbolic links, to learn the target.

=cut

sub readlink {
    my $self  = shift;
    my $vfile = shift;
    my $file  = $self->find_file($vfile);

    $self->log(debug => "readlink(%s)", $file);

    return 0 unless($file); # such file or directory
    return readlink $file;
}

=head2 getdir

 @files = $self->getdir($virtual_path);

Returns a list of filenames from the virtual dir.

This is used to obtain directory listings. Its opendir(), readdir(),
filldir() and closedir() all in one call.

=cut

sub getdir {
    my $self = shift;
    my $vdir = shift;
    my $dir  = $self->find_file($vdir);
    my($DH, @files);

    $self->log(debug => "getdir(%s)", $dir);

    opendir($DH, $dir) or return -&POSIX::ENOENT;
    @files = readdir $DH;

    return @files, 0;
}

=head2 mknod

 $errno = $self->mknod($virtual_path);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

This function is called for all non-directory, non-symlink nodes, not
just devices.

=cut

sub mknod {
    my $self = shift;
    $self->log(error => "mknod => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 mkdir

 $errno = $self->mkdir($virtual_path, $mode);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to create a directory.

=cut

sub mkdir {
    my $self = shift;
    $self->log(error => "mkdir => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 unlink

 $errno = $self->unlink($virtual_path);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to remove a file, device, or symlink.

=cut

sub unlink {
    my $self = shift;
    $self->log(error => "unlink => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 rmdir

 $errno = $self->rmdir($virtual_path);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to remove a directory.

=cut

sub rmdir {
    my $self = shift;
    $self->log(error => "rmdir => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 symlink

 $errno = $self->symlink($virtual_path, $symlink_name);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to create a symbolic link.

=cut

sub symlink {
    my $self = shift;
    $self->log(error => "symlink => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 rename

 $errno = $self->rmdir($old_virtual_path, $new_virtual_path);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to rename a file, and/or move a file from one directory to another.

=cut

sub rename {
    my $self = shift;
    $self->log(error => "rename => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 link

 $errno = $self->link($virtual_path, $hardlink_name);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to create hard links.

=cut

sub link {
    my $self = shift;
    $self->log(error => "link => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 chmod

 $errno = $self->chmod($virtual_path, $mode);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to change permissions on a file/directory/device/symlink.

=cut

sub chmod {
    my $self = shift;
    $self->log(error => "chmod => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 chown

 $errno = $self->chown($virtual_path, $uid, $gid);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to change ownership of a file/directory/device/symlink.

=cut

sub chown {
    my $self = shift;
    $self->log(error => "chown => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 truncate

 $errno = $self->truncate($virtual_path, $offset);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented. 

Called to truncate a file, at the given offset.

=cut

sub truncate {
    my $self = shift;
    $self->log(error => "truncate => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 utime

 $errno = $self->utime($virtual_path, $actime, $modtime);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called to change access/modification times for a
file/directory/device/symlink.

=cut

sub utime {
    my $self = shift;
    $self->log(error => "utime => ro");
    return -&POSIX::EROFS; # cannot write
}

=head2 open

 $errno = $self->open($virtual_path, $mode);

Called to open a file.

=cut

sub open {
    my $self  = shift;
    my $vfile = shift;
    my $mode  = shift;
    my $file  = $self->find_file($vfile);

    $self->log(debug => 'open(%s, %s)', $file, $mode);

    return -$! unless(sysopen my $FH, $file, $mode);
    return 0;
}

=head2 read

 $ret = $self->read($virtual_path, $buf_size, $offset);

C<$ret> can be either C<$errno> or data read.

Called in an attempt to fetch a portion of the file.

=cut

sub read {
    my $self    = shift;
    my $vfile   = shift;
    my $bufsize = shift;
    my $offset  = shift;
    my $file    = $self->find_file($vfile);
    my $size    = -s $file;
    my($FH, $buf);

    $self->log(debug => 'read(%s, %i, %i)', $file, $bufsize, $offset);

    # check for file existence
    return -&POSIX::ENOENT() unless(defined $size);

    # open and read file
    sysopen $FH, $file, &POSIX::O_RDONLY   or return -&POSIX::ENOSYS;
    sysseek $FH, $offset, &POSIX::SEEK_SET or return -&POSIX::ENOSYS;
    sysread $FH, $buf, $bufsize    or return -&POSIX::ENOSYS;

    return $buf;
}

=head2 write

 $ret = $self->write($virtual_path, $buf_size, $offset);

C<$ret> can be either C<$errno> or length of data read.

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

Called in an attempt to write (or overwrite) a portion of the file.

=cut

# Be prepared because $buffer could contain random binary data with
# NULLs and all sorts of other wonderful stuff.
sub write {
    my $self = shift;
    $self->log(error => 'write => ro');
    return -&POSIX::EROFS;
}

=head2 statfs

 @ret = $self->statfs;

C<@ret> can be on of:

 ($namelen, $files, $files_free, $blocks, $blocks_avail, $blocksize)
 -ENOANO()

=cut

sub statfs {
    my $self  = shift;
    my $total = {
        blocks => 0, # total blocks
        bfree  => 0, # total free blocks
        files  => 0, # total inodes
        ffree  => 0, # total free inodes
    };

    $self->log(debug => "statfs()");

    return(
        255,
        $total->{'blocks'},
        $total->{'bfree'},
        $total->{'files'},
        $total->{'ffree'},
        $self->root->block_size,
    );
}

=head2 flush

 $errno = $self->flush($virtual_path);

Called to synchronise any cached data. This is called before the file
is closed. It may be called multiple times before a file is closed.

=cut

sub flush {
    my $self  = shift;
    my $vfile = shift;
    $self->log(debug => 'flush...');
    return 0;
}

=head2 release

 $errno = $self->release($virtual_path, $flags);

Called to indicate that there are no more references to the file.
Called once for every file with the same pathname and flags as were
passed to open.

=cut

sub release {
    my $self  = shift;
    my $vfile = shift;
    my $flags = shift;
    $self->log(debug => 'release...');
    return 0;
}

=head2 fsync

 $int = $self->fsync($virtual_path, $flags);

Called to synchronise the file's contents. If flags is non-zero, only
synchronise the user data. Otherwise synchronise the user and meta data.

=cut

sub fsync {
    my $self  = shift;
    my $vfile = shift;
    my $flags = shift;
    $self->log(debug => 'fsync...');
    return 1;
}

=head2 setxattr

 $errno = $self->setxattr(
              $virtual_path, $attr_name, $attr_value, $flags
          );

Called to set the value of the named extended attribute.

Will always return C<-EOPNOTSUPP()> since write is not supported.

=cut

sub setxattr {
    my $self  = shift;
    my $vfile = shift;
    my $name  = shift;
    my $value = shift;
    my $flags = shift; # OR-ing of XATTR_CREATE and XATTR_REPLACE

    $self->log(debug => 'setxattr(%s, %s, %s, %s)',
        $vfile, $name, $value, $flags,
    );

    return -&POSIX::EOPNOTSUPP();
}

=head2 getxattr

 $ret = $self->getxattr($virtual_path, $attr_name);

Called to get the value of the named extended attribute.

=cut

# If flags is set to XATTR_CREATE and the extended attribute already exists,
# this should fail with -EEXIST(). If flags is set to XATTR_REPLACE and the
# extended attribute doesn't exist, this should fail with -ENOATTR().
sub getxattr {
    my $self  = shift;
    my $vfile = shift;
    my $name  = shift;
    my $value;

    $self->log(debug => 'getxattr(%s, %s)', $vfile, $name);
 
    return $value if($value);
    return 0; # no such name
}

=head2 listxattr

 @list = $self->listxattr;

Called to get the value of the named extended attribute.

=cut

sub listxattr {
    my $self = shift;
    $self->log(debug => 'listxattr...');
    return 0;
}

=head2 removexattr

 $errno = $self->removexattr($virtual_path, $attr_name);

This method will always return C<-&POSIX::EROFS>, since write support is not
implemented.

=cut

sub removexattr {
    my $self  = shift;
    my $vfile = shift;
    my $name  = shift;
    $self->log(debug => 'removexattr => ro');
    return -&POSIX::EROFS; # cannot write
}

=head1 AUTHOR

See L<Fuse::Template>

=cut

1;


lib/Fuse/Template/Root.pm
package Fuse::Template::Root;

=head1 NAME

Fuse::Template::Root - Root filesystem object

=head1 SYNOPSIS

    package Class;
    # will import RootObject typeconstraint
    use Fuse::Template::Root qw/RootObject/;
    has foo => ( isa => RootObject, coerce => 1 );
    ...

    package main;

    # same result:
    my $obj = Class->new( foo => { path => $some_root } );
    my $obj = Class->new( foo => $some_root );
    my $obj = Class->new(
                  foo => Fuse::Template::Root->new({
                      path => $some_root
                  }),
              );

=head1 DESCRIPTION

The C<RootObject> can coerce either Str and HashRef into a
L<Fuse::Template::Root> object. The HashRef will be used as object
constructor, while the Str will be set as L</path> attribute.

=cut

use Moose;
use MooseX::Types -declare => [qw/RootObject/];
use MooseX::Types::Moose qw(:all);
use overload q("") => sub { shift->path }, fallback => 1;

subtype RootObject, as Object;
coerce RootObject, (
    from Str,     via { Fuse::Template::Root->new(path => $_) },
    from HashRef, via { Fuse::Template::Root->new($_) },
);

=head1 ATTRIBUTES

=head2 path

Holds a string representing the path to where the template files
exist. This attribute is required.

=cut

has path => (
    is => 'ro',
    isa => 'Str',
    required => 1,
);

=head2 mode

Holds the file mode for the target mount directory. Default value
is C<0775>. (octal value)

=cut

has mode => (
    is => 'ro',
    isa => 'Int',
    default => 0775,
);

around mode => sub {
    my $next = shift;
    my $self = shift;

    return @_ ? $self->$next((0040 << 9) + $_[0]) : $self->$next;
};

=head2 uid

The user id of who owns the target mount path. Defaults to the
C<$REAL_USER_ID>

=cut

has uid => (
    is => 'ro',
    isa => 'Int',
    default => $<,
);

=head2 gid

The group id of who owns the target mount path. Defaults to
the first integer from C<$REAL_GROUP_ID>.

=cut

has gid => (
    is => 'ro',
    isa => 'Int',
    default => sub { $( =~ /(\d+)/ },
);

=head2 ctime

Last change time of the target mount path. Defaults to
C<time> when the object got created.

=cut

has ctime => (
    is => 'ro',
    isa => 'Int',
    default => time,
);

=head2 mtime

Last modify time of the target mount path. Defaults to
C<time> when the object got created.

=cut

has mtime => (
    is => 'ro',
    isa => 'Int',
    default => time,
);

=head2 atime

Last access time of the target mount path. Defaults to
C<time> when the object got created.

=cut

has atime => (
    is => 'ro',
    isa => 'Int',
    default => time,
);

=head2 block_size

Preferred block size for file system I/O. Defaults to 1024.
(The default value might change)

=cut

has block_size => (
    is => 'ro',
    isa => 'Int',
    default => 1024,
);

=head1 AUTHOR

See L<Fuse::Template>

=cut

1;


lib/Fuse/Template/Schema.pm
package Fuse::Template::Schema;

=head1 NAME

Fuse::Template::Schema - Dynamic load of database schema

=head1 DESCRIPTION

Using L<DBIx::Class::Schema::Loader> to read table definitions from
database will use the default moniker rules to name resultsets from
table names. See L<DBIx::Class::Schema::Loader::Base/moniker_map> for
details.

=head1 SYNOPSIS

 use Fuse::Template::Schema qw/Schema/;

 # will import Schema typeconstraint
 use Fuse::Template::Root qw/Schema/;
 has foo => ( isa => Schema, coerce => 1 );
 ...

=cut

BEGIN {
    # required for extends + MooseX::Types to work properly
    use Moose;
    extends qw/DBIx::Class::Schema::Loader/;
}

use MooseX::Types -declare => [qw/Schema/];
use MooseX::Types::Moose qw(:all);
use DBIx::Class;

subtype Schema, as Object;
coerce Schema, (
    from Str,     via { &from_string },
    from HashRef, via { &from_hashref },
);

unless($ENV{'SCHEMA_LOADER_BACKCOMPAT'}) {
    __PACKAGE__->loader_options(
        naming => 'v7',
        use_namespaces => 1,
    );
}

=head1 FUNCTIONS

=head2 from_string

 $schema = from_string("$schema_class $dsn");
 $schema = from_string("$schema_class $dsn $username $password");
 $schema = from_string("$dsn ...");

=cut

sub from_string {
    my $input = $_;

    if($input =~ /^(dbi:.*)/) {
        return __PACKAGE__->connect(split /\s+/, $input);
    }
    elsif($input =~ /([\w:]+)\s(.*)/) {
        my $class = $1;
        my $dsn   = $2; # "$dsn $username $password";
        if(eval "require $class") {
            return $class->connect(split /\s+/, $dsn);
        }
    }
    else {
        confess "invalid arguments";
   }
}

=head2 from_hashref

 $schema = from_hashref({
               schema => $class_name, # optional
               dsn => $dbi_dsn,
               username => $str,
               password => $str,
               %dbi_params, # optional
           });

=cut

sub from_hashref {
    my $args   = $_;
    my $schema = delete $args->{'schema'};

    confess "'dsn' is required" unless($args->{'dsn'});

    $schema ||= __PACKAGE__;

    return $schema->connect(
        delete $args->{'dsn'},
        delete $args->{'username'},
        delete $args->{'password'},
        $args,
    );
}

=head1 AUTHOR

See L<Fuse::Template>

=cut

1;


lib/Fuse/Template.pm
package Fuse::Template;

=head1 NAME

Fuse::Template - Mount a directory with templates.

=head1 VERSION

0.0201

=head1 USAGE

C<sudo> might be optional.

=head2 MOUNT

Example SQLite:

 $ sudo fuse-template
   --root /path/to/templates \
   --mountpoint /path/to/mountpoint \
   --schema dbi:SQLite:path/to/db \
   ;

Example MySQL:

 $ sudo fuse-template
   --root /path/to/templates \
   --mountpoint /path/to/mountpoint \
   --schema "dbi:mysql:database=MyDB;host=127.1;port=3006 user pass" \
   ;

See L<DBI> for possible C<--schema> formats.

=head2 UNMOUNT

This needs to be done manually for now. (Patches are welcome!)

 sudo umount /path/to/mountpoint;

=head1 DESCRIPTION

IMPORTANT! Version 0.02 of L<Fuse::Template> breaks backward compat by
using the new "v7" naming convention from L<DBIx::Class::Schema::Loader>.
Set the C<SCHEMA_LOADER_BACKCOMPAT> environment variable, if you want the
old behavior.

The idea with this project is to auto-maintain password files, and other
plain text files for different systems, with data from a database.

See C<examples/> in this distribution for templates.

Example file system structure:

 root/               -> mountpooit/
 root/somefile       -> mountpoint/somefile
 root/foo.tt         -> mountpoint/foo
 root/bar/baz.txt.tt -> mountpoint/bar/baz.txt

C<somefile> is accessible directly at C<mountpoint>. Template files (.tt)
are parsed and the output is accessible on mountpoint side. This is done
using L<Template::Toolkit>.

Files in mountpoint are read-only for now.

=head1 TEMPLATES

Here is an example template:

 root = [% self.root %]
 [% rs = schema.resultset('MyTable') %]
 [% rs = MyTable; # same as schema.resultset(...) %]
 [% WHILE (row = rs.next) # get next row in resultset %]
 col_name = [% row.col_name # retrieve column data %]
 [% END %]

See L<DBIx::Class> for information on how to use the L<schema> object, and
L<DBIx::Class::Resultset> on how to select data and more...

Available variables:

 root
 mountpoint
 mountopts
 self # this object
 schema # DBIx::Class object
 + all resultsetset/tables

Resultset use this naming convention to convert tablenames:

    Table Name  | Moniker Name
    ---------------------------
    luser       | Luser
    luser_group | LuserGroup
    luser-opts  | LuserOpts

See L<DBIx::Class::Schema::Loader/moniker_map> for details

=head1 INSTALLATION

On ubuntu:

    $ sudo aptitude install libfuse-perl
    $ wget -q http://xrl.us/cpanm | perl - --sudo Fuse::Template

See L<App::cpanminus> if you are curious about the "wget" command.

=cut

use Moose;
use Fuse::Template::Schema qw/Schema/;
use Fuse::Template::TT;

with qw/Fuse::Template::Sys/;

our $VERSION = '0.0201';

=head1 ATTRIBUTES

=head2 root

 $object = $self->root;

C<$object> must be stringified to path where templates/files are located.
C<$object> must follow the same API as L<Fuse::Template::Root>.

Can be set in constructor as:

 root => \%constructor_args
 root => $path_to_root

C<%constructor_args> can override defaults, while C<$path_to_root> will
only override C<path> attribute in root-object.

=head2 mountpoint

 $path = $self->mountpoint;

Path to where the filesystem should be mounted.

=cut

has mountpoint => (
    is => 'ro',
    isa => 'Str',
    documentation => 'Path to mount point',
    #required => 1,
);

=head2 mountopts

 $str = $self->mountopts;

Example mount option: "allow_other"

=cut

has mountopts => (
    is => 'ro',
    isa => 'Str',
    documentation => 'Mount options. Example: "allow_other"',
);

=head2 schema

 $dbic_object = $self->schema;

Can be set in constructor as:

 schema => \%args
 schema => "$dsn $username $password"

See L<Fuse::Template::Schema> for details.

=cut

has schema => (
    is => 'ro',
    isa => Schema,
    coerce => 1,
    lazy_build => 1,
    documentation => 'Schema name or dsn',
);

sub _build_schema {
    confess "Cannot build schema without information!";
}

=head2 debug

 $bool = $self->debug;

Enable/disable debug output from L<Fuse>. Default is disabled.

=cut

has debug => (
    is => 'ro',
    isa => 'Bool',
    default => 0,
    documentation => 'Enable/disable debug output',
);

has _template => (
    is => 'ro',
    isa => 'Object',
    handles => [qw/render/],
    lazy_build => 1,
);

sub _build__template {
    my $self = shift;

    return Fuse::Template::TT->new(
        paths => [$self->root->path],
        vars => {
            root => $self->root->path,
            mountpoint => $self->mountpoint,
            mountopts => $self->mountopts,
            self => $self,
            ft => $self,
        },
    );
}

=head1 METHODS

=head2 find_file

 $real_path = $self->find_file($virtual_path);

Returns path to the file in root path.

=cut

sub find_file {
    my $self  = shift;
    my $vfile = shift;
    my $root  = $self->root->path;
    my $file;

    $root  =~ s,/+$,,;
    $vfile =~ s,^/+,,;

    if(-e "$root/$vfile.tt") {
        $file = "$root/$vfile.tt";
    }
    elsif(-e "$root/$vfile") {
        $file = "$root/$vfile";
    }

    $self->log(debug => "%s = %s + find_file(%s)", $file, $root, $vfile);

    return $file;
}

=head2 log

 $bool = $self->log($level => $format => @args);

=cut

sub log {
    return unless $_[0]->debug;

    my $self   = shift;
    my $level  = shift;
    my $format = shift;
    my @args;

    for(@_) {
        push @args, defined $_ ? $_ : '__UNDEF__';
    }

    warn sprintf "$level - $format\n", @args;
}

# around ::Sys::getdir(...);
around getdir => sub {
    my $next  = shift;
    my $self  = shift;
    my @files = $self->$next(@_);

    return map { s/\.tt$//; $_; } @files;
};

# around ::Sys::read(...);
around read => sub {
    my($next, $self, $vfile, $bufsize, $offset) = @_;
    my $file = $self->find_file($vfile);

    # standard file
    if($file =~ /\.tt$/) {
        $vfile =~ s,^/,,;

        if($self->has_schema) {
            my $vars = $self->_template->vars;
            my $schema = $self->schema;
            $vars->{'schema'} = $schema;
            $vars->{$_} = $schema->resultset($_) for($schema->sources);
        }

        my $output = $self->render("$vfile.tt");
        return substr $output, $offset, $bufsize;
    }
    else {
        return $self->$next($vfile, $bufsize, $offset);
    }
};

=head1 AUTHOR

Jan Henning Thorsen

=head1 LICENSE

This library is free software, you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;
